### PR TITLE
Parse v1 roots

### PR DESCRIPTION
**Pull request by ctz**
This should address issue #13.

### CODE REVIEW

**Comment by ctz** [comment]
Apologies, this PR unfortunately relies on the other too.

File: tests/integration.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  &ECDSA_P384_SHA512,
+  &RSA_PKCS1_2048_8192_SHA1,
+  &RSA_PKCS1_2048_8192_SHA256,
+  &RSA_PKCS1_2048_8192_SHA384,
+  &RSA_PKCS1_2048_8192_SHA512,
+  &RSA_PKCS1_3072_8192_SHA384
+];
+
+fn read_bytes(filename: &str) -> Vec<u8> {
+  use std::fs::File;
```
Comment (briansmith) [inline tests/integration.rs]: Although I don't like it much, I decided to use the 4-space indent in the rest of the project since that seems to be the standard in Rust. So, please use four-space indent.

File: tests/integration.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    .expect("cannot read file");
+  out
+}
+
+#[test]
+pub fn netflix()
+{
+  let ee = read_bytes("tests/netflix/ee.der");
+  let inter = read_bytes("tests/netflix/inter.der");
+  let ca = read_bytes("tests/netflix/ca.der");
```
Comment (briansmith) [inline tests/integration.rs]: Use `include_bytes!` here. Then we don't need to depend on `std::fs`.

File: src/cert.rs [hunk 3]
```diff
@@ -37,6 +37,78 @@ pub struct Cert<'a> {
     pub subject_alt_name: Option<untrusted::Input<'a>>,
 }
 
+pub fn parse_root_cert_v1<'a>(cert_der: untrusted::Input<'a>)
```
Comment (briansmith) [inline src/cert.rs]: I would rather keep this in `trust_anchor_util`. Also, I think that this should just return a `TrustAnchor` directly, instead of constructing a `Cert` first. I think that would simplify it quite a bit. What do you think?

**Comment by briansmith** [comment]
Are you the (sole) copyright holder(s) for your changes? If not, who owns the copyrights?

In your commits, please add the license statement at https://github.com/briansmith/ring/#contributing.

**Comment by ctz** [comment]
That latest push should address your earlier comments:

- 4-space indentation everywhere
- using `include_bytes!`. I hadn't actually come across this before. Very useful!
- moving v1 parsing to `trust_anchor_util` and simplify it by going straight to a `TrustAnchor`

I've also put the licensing notice in commits.

File: tests/integration.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
+// Permission to use, copy, modify, and/or distribute this software for any
```
Comment (briansmith) [inline tests/integration.rs]: Please add the copyright line. (Whoever the copyright holder is.)

File: src/trust_anchor_util.rs [hunk 5]
```diff
@@ -52,3 +60,55 @@ pub fn generate_code_for_trust_anchors(name: &str,
...
+     *
+     * CertificateInfo ::= SEQUENCE {
+     *   version [0] Version DEFAULT v1988,
+     *   serialNumber CertificateSerialNumber,
+     *   signature AlgorithmIdentifier,
+     *   issuer Name,
+     *   validity Validity,
+     *   subject Name,
+     *   subjectPublicKeyInfo SubjectPublicKeyInfo }
+     */
```
Comment (briansmith) [inline src/trust_anchor_util.rs]: Rather than copy the documentation from the RFC, just cite this documentation with a link.

We copy/pasted the ASN.1 definitions into the source code for mozilla::pkix, but for webpki I decided that was a mistake; it's better to read them out of the source document if we want to read them at all. That way we don't have to worry about whether we copy/pasted the correct thing and/or copy/pasted wrongly. In this case, https://tools.ietf.org/html/rfc5280#section-4.1 is a good link.

File: src/trust_anchor_util.rs [hunk 6]
```diff
@@ -52,3 +60,55 @@ pub fn generate_code_for_trust_anchors(name: &str,
...
+                try!(certificate_serial_number(inf));
+                try!(der::expect_tag_and_get_input(inf, der::Tag::Sequence)); // sigalg
+                try!(der::expect_tag_and_get_input(inf, der::Tag::Sequence)); // issuer
+                try!(der::expect_tag_and_get_input(inf, der::Tag::Sequence)); // validity
+                let subject =
+                    try!(der::expect_tag_and_get_input(inf, der::Tag::Sequence));
+                let spki =
+                    try!(der::expect_tag_and_get_input(inf, der::Tag::Sequence));
+
+                let anchor: TrustAnchor<'a> = TrustAnchor {
```
Comment (briansmith) [inline src/trust_anchor_util.rs]: Is the type annotation necessary here? I would prefer just:

```Ok(TrustAnchor {
...
})```

File: src/trust_anchor_util.rs [hunk 7]
```diff
@@ -16,25 +16,33 @@
 
 use std;
 use super::{Error, TrustAnchor};
-use super::cert::{EndEntityOrCA, parse_cert};
+use super::cert::{EndEntityOrCA, Cert, parse_cert};
+use super::cert::certificate_serial_number;
+use super::der;
 use untrusted;
 
+fn trust_anchor_from_cert<'a>(cert: Cert<'a>) -> TrustAnchor<'a> {
```
Comment (briansmith) [inline src/trust_anchor_util.rs]: Please put this function between `cert_der_as_trust_anchor` and `parse_cert_v1`. In general, we put public things ahead of the implementation details of the public thing, and we try to put things roughly in order that they are used, so that somebody wanting to read `cert_der_as_trust_anchor` should be able to grep for it and then scroll down.

File: src/trust_anchor_util.rs [hunk 8]
```diff
@@ -16,25 +16,33 @@
...
-    Ok(TrustAnchor {
-        subject: cert.subject.as_slice_less_safe(),
-        spki: cert.spki.as_slice_less_safe(),
-        name_constraints: cert.name_constraints
-                              .map(|nc| nc.as_slice_less_safe())
-    })
+    parse_cert(cert_der, EndEntityOrCA::EndEntity)
+        .map(trust_anchor_from_cert)
+        .or_else(|err| parse_cert_v1(cert_der)
+                 .or(Err(err)))
```
Comment (briansmith) [inline src/trust_anchor_util.rs]: webpki was the very first Rust code I wrote, and so I wrote it more like C++. In particular, I didn't use this chaining style like this. Looking at this, it doesn't seem so bad. But, I will need a little time to read through the existing code and see how we should do this.

Comment (briansmith) [inline src/trust_anchor_util.rs]: Given the discussion in #17, it seems like we should just use the new parser for all trust anchors. That would simplify the code in this PR. It won't completely solve #17, but fixing #17 would be just one step away.

**Comment by briansmith** [comment]
@ctz: Please check out https://github.com/briansmith/webpki/commits/more-liberal-trust-anchor-parsing, which makes some tweaks to your submission here. Let me know what you think. Thanks!

Also, I didn't realize that you had addressed some of the previous feedback. :( I had been waiting for a response from you. It seems I'm no good at noticing GitHub notifications. Sorry for the huge latency.

**Comment by briansmith** [comment]
Rebased:  https://github.com/briansmith/webpki/commits/more-liberal-trust-anchor-parsing,

**Comment by ctz** [comment]
LGTM! (and rustls passes tests against the tip of that branch). If you can merge that then I can point rustls at crates.io and get rid of my forks :)

**Comment by briansmith** [comment]
Can you please submit a PR with your test case?

**Comment by briansmith** [comment]
Thanks! This was fixed in 0.2.1:

220acd1 Bump version number for publication.
8fd9a89 Add tests for -ve and zero serials in roots
8d33099 Allow trust anchors with invalid serial numbers.
591a11b Clarify error handling in cert_der_as_trust_anchor.
8980851 Fix warnings about unused results in trust_anchor_util.rs.
d4d2886 Parse v1 implicit certs for trust_anchor_util.
